cmd/snap-bootstrap/triggerwatch/triggerwatch.go

```diff
@@ -71,17 +112,40 @@ func Wait(timeout time.Duration) error {
 		go dev.WaitForTrigger(detectKeyCh)
 		defer dev.Close()
 	}
+	foundDevice := len(devices) != 0
 
-	select {
-	case kev := <-detectKeyCh:
-		if kev.Err != nil {
-			return kev.Err
+	start := time.Now()
+	for {
+		timePassed := time.Now().Sub(start)
+		relTimeout := timeout - timePassed
+		relDeviceTimeout := deviceTimeout - timePassed
+		select {
+		case kev := <-detectKeyCh:
+			if kev.Err != nil {
+				return kev.Err
+			}
+			// channel got closed without an error
+			logger.Noticef("%s: + got trigger key %v", kev.Dev, triggerFilter.Key)
+			return nil
+		case <-time.After(relTimeout):
+			return ErrTriggerNotDetected
+		case <-time.After(relDeviceTimeout):
+			if !foundDevice {
+				return ErrNoMatchingInputDevices
+			}
+		case uevent := <-ueventQueue:
+			dev, err := trigger.Open(triggerFilter, uevent.Env["DEVNAME"])
+			if dev != nil && err == nil {
+				foundDevice = true
+				defer dev.Close()
+				go dev.WaitForTrigger(detectKeyCh)
+			}
+		case <-sigs:
+			logger.Noticef("Switching root")
+			err := syscall.Chroot("/sysroot")
+			if err != nil {
+				return errors.New("Cannot change root")
```

Comment:
I'd also include the original error to this line, maybe it can contain some important hints.